core/src/main/java/org/bitcoinj/wallet/Wallet.java

```diff
@@ -2214,11 +2216,13 @@ private void processTxFromBestChain(Transaction tx, boolean forceAddToPool) thro
                 // no need to addTransactionsDependingOn(doubleSpendTxns) because killTxns() already kills dependencies;
                 killTxns(doubleSpendTxns, tx);
             }
-            // disconnect irrelevant inputs (otherwise might cause serialization issue)
-            for(TransactionInput input : tx.getInputs()) {
-                TransactionOutput output = input.getOutpoint().getConnectedOutput();
-                if(output != null && !output.isMineOrWatched(this)) {
-                    input.disconnect();
+            if(!hasOutputsToMe && !hasOutputsFromMe && !forceAddToPool) {
+                // disconnect irrelevant inputs (otherwise might cause serialization issue)
+                for (TransactionInput input : tx.getInputs()) {
+                    TransactionOutput output = input.getConnectedOutput();
+                    if(output != null && !output.isMineOrWatched(this)) {
```

Comment:
Apparently my home setup failed me and I didn't notice that test fail because of missing `!`